tests/41end-to-end-keys/04-query-key-federation.pl

```diff
@@ -28,8 +28,11 @@
 
          # TODO: Check that the content matches what we uploaded.
 
-         assert_eq( $alice_device_keys->{"unsigned"}->{"device_display_name"},
-                    "test display name" );
+         # Device display names are optional for POST /user/keys/query responses.
+         # If one exists, ensure it's the one we expected.
+         my $device_display_name = $alice_device_keys->{"unsigned"}->{"device_display_name"};
+         (!defined $device_display_name) or ($device_display_name == "test display name") or
+            croak "Unexpected device_display_name: $device_display_name";
```

Comment:
FWIW I have no idea if this is idomatic perl (from #1307).

IIRC I manually inlined `assert_eq` to get the `a == b or croak "wrong"` form, and then shoved an extra defined check at the start.

Reply:
I'd say in this situation it would be more idiomatic to use `||` between the conditions:

```
         (!defined $device_display_name || $device_display_name eq "test display name") or
            croak "Unexpected device_display_name: $device_display_name";
```

Note also the use of `eq` instead of `==`: the latter is for *numeric* comparison:

```
$ perl -e '"a" == "b" and print "true\n";'
true
```

A simpler approach here would be to do something like:
```
         my $device_display_name = $alice_device_keys->{"unsigned"}->{"device_display_name"} // "";
         assert_eq "test display name", $device_display_name, "device display name";
```

(`//` is like an `||` operator, but whereas `||` returns its right-hand operand if the left-hand one is falsey, `//` only returns its right-hand operand if the left-hand operand is `undef`. See https://perldoc.perl.org/perlop#Logical-Defined-Or)

Reply:
> Note also the use of eq instead of ==: the latter is for numeric comparison:

(and yes, this is a massive footgun)

Reply:
Ahh, yes, `assert_eq` does indeed use `eq`.

https://github.com/matrix-org/sytest/blob/32129d98c33ff85e6e73e67147ea3e4cd04f630d/lib/SyTest/Assertions.pm#L61-L67

Presumably this is bogus with `==` then. Can you fix that up here Andrew---and the test from #1307?

Reply:
Sure, done in ad8ad254.

Reply:
I actually wonder if my original code would be easier to read for the average programmer?

```perl
         if (exists( $alice_device_keys->{"unsigned"}->{"device_display_name"} )) {
            assert_eq(
               $alice_device_keys->{"unsigned"}->{"device_display_name"},
               "test display name",
            );
         }
```

but either way I'm happy as long as it works.